### ISSUE TITLE
connect to the socket ”/var/run/audispd_events“, status is all  CONNECTING, no CONNECTED?

### ISSUE THREAD

**Issue by stillyger** [linux-audit/audit-userspace#300]
hi,  i find a problem.

The steps are as follows:
1. enable the af_unix plug-in；
2. run test process, it will connect the socket file ”/var/run/audispd_events“,  restart the test process at an interval of 5s；
3. after 1-2 hours，will see the problem, the connect status is all ”CONNECTING“, no ”CONNECTED“
```
[root@k8s-single audit]# netstat -anp | grep audi
unix  2      [ ACC ]     STREAM     LISTENING     390679   49098/auditd         /var/run/audispd_events
unix  2      [ ]         STREAM     CONNECTING    0        -                    /var/run/audispd_events
unix  3      [ ]         STREAM     CONNECTED     390682   49098/auditd
unix  2      [ ]         STREAM     CONNECTING    0        -                    /var/run/audispd_events
unix  2      [ ]         STREAM     CONNECTING    0        -                    /var/run/audispd_events
unix  2      [ ]         DGRAM                    390680   49098/auditd
unix  2      [ ]         STREAM     CONNECTING    0        -                    /var/run/audispd_events
unix  2      [ ]         STREAM     CONNECTING    0        -                    /var/run/audispd_events
unix  2      [ ]         STREAM     CONNECTING    0        -                    /var/run/audispd_events
unix  3      [ ]         STREAM     CONNECTED     390681   49098/auditd
[root@k8s-single audit]#
```

/etc/audit/plugins.d/af_unix.conf:

```
active = yes
direction = out
path = builtin_af_unix
type = builtin
args = 0640 /var/run/audispd_events
format = string
```

/etc/audit/auditd.conf:

```
local_events = yes
write_logs = yes
log_file = /var/log/audit/audit.log
log_group = root
log_format = ENRICHED
flush = INCREMENTAL_ASYNC
freq = 50
max_log_file = 8
num_logs = 5
priority_boost = 4
name_format = NONE
##name = mydomain
max_log_file_action = ROTATE
space_left = 75
space_left_action = SYSLOG
verify_email = yes
action_mail_acct = root
admin_space_left = 50
admin_space_left_action = SUSPEND
disk_full_action = SUSPEND
disk_error_action = SUSPEND
use_libwrap = yes
##tcp_listen_port = 60
tcp_listen_queue = 5
tcp_max_per_addr = 1
##tcp_client_ports = 1024-65535
tcp_client_max_idle = 0
transport = TCP
krb5_principal = auditd
##krb5_key_file = /etc/audit/audit.key
distribute_network = no
q_depth = 400
overflow_action = SYSLOG
max_restarts = 10
plugin_dir = /etc/audit/plugins.d
```

```
top - 16:48:06 up  7:24,  4 users,  load average: 0.38, 0.44, 0.83
Threads:   3 total,   0 running,   2 sleeping,   1 stopped,   0 zombie
%Cpu(s):  0.0 us,  3.3 sy,  0.0 ni, 96.7 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem :   3379.2 total,    469.6 free,   1257.8 used,   1651.8 buff/cache
MiB Swap:      0.0 total,      0.0 free,      0.0 used.   1896.9 avail Mem

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
  49098 root      16  -4   95520   4876   1572 S   0.0   0.1  58:52.23 auditd
  49099 root      16  -4   95520   4876   1572 S   0.0   0.1   2:41.19 auditd
  49100 root      16  -4   95520   4876   1572 t   0.0   0.1  19:25.61 auditd
```

```
[root@k8s-single audit]# strace -p 49100
strace: Process 49100 attached
futex(0xaaaad2b242e0, FUTEX_WAKE_PRIVATE, 1) = 0
futex(0xffff923f3a78, FUTEX_WAKE_PRIVATE, 1) = 1
futex(0xffff923f3a78, FUTEX_WAKE_PRIVATE, 1) = 0
futex(0xaaaad2b24338, FUTEX_WAIT_BITSET_PRIVATE|FUTEX_CLOCK_REALTIME, 0, NULL, FUTEX_BITSET_MATCH_ANY) = -1 EAGAIN (Resource temporarily unavailable)
futex(0xaaaad2b242e0, FUTEX_WAKE_PRIVATE, 1) = 0
futex(0xaaaad2b2433c, FUTEX_WAIT_BITSET_PRIVATE|FUTEX_CLOCK_REALTIME, 0, NULL, FUTEX_BITSET_MATCH_ANY) = -1 EAGAIN (Resource temporarily unavailable)
futex(0xaaaad2b242e0, FUTEX_WAKE_PRIVATE, 1) = 0
futex(0xffff923f3a78, FUTEX_WAKE_PRIVATE, 1) = 0
futex(0xaaaad2b24338, FUTEX_WAIT_BITSET_PRIVATE|FUTEX_CLOCK_REALTIME, 0, NULL, FUTEX_BITSET_MATCH_ANY) = -1 EAGAIN (Resource temporarily unavailable)
futex(0xaaaad2b242e0, FUTEX_WAKE_PRIVATE, 1) = 0
futex(0xaaaad2b2433c, FUTEX_WAIT_BITSET_PRIVATE|FUTEX_CLOCK_REALTIME, 0, NULL, FUTEX_BITSET_MATCH_ANY) = -1 EAGAIN (Resource temporarily unavailable)
futex(0xaaaad2b242e0, FUTEX_WAKE_PRIVATE, 1) = 0
futex(0xffff923f3a78, FUTEX_WAIT_PRIVATE, 2, NULL) = -1 EAGAIN (Resource temporarily unavailable)
futex(0xffff923f3a78, FUTEX_WAKE_PRIVATE, 1) = 0
futex(0xffff923f3a78, FUTEX_WAIT_PRIVATE, 2, NULL) = -1 EAGAIN (Resource temporarily unavailable)
futex(0xffff923f3a78, FUTEX_WAKE_PRIVATE, 1) = 0
futex(0xffff923f3a78, FUTEX_WAKE_PRIVATE, 1) = 0

```

**Comment by stillyger** [comment]
continue analysis,
when client connet failed,  and listen queue is full,  all connet status are connecting.

function call：
send_af_unix_string 
   -> rc = write(conn, s, len);    => return failed, rc=-1, errno=EPIPE;
   ->conn=-1                           => conn  never change,  always -1;
   ->stop_watching
   ->start_watching                 
       --> af_unix_accept           => in abnormal scenarios ,  accept never called

suspecte that there may be some error in ev_io_start?

**Comment by stillyger** [comment]
We found that:
The following global variables, There may be two threads working at the same time.
fdchangecnt，
fdchanges:   

```
thread 1:main process
auditd -> loop -> ev_run -> fd_reify :  (first get fdchangecnt value: int changecnt = fdchangecnt;  in the end set value, fdchangecnt -= changecnt;)

thread2: audisp process
audisp -> outbound_thread_main-> event_loop -> send_af_unix_string -> start_watching -> ev_io_start -> fd_change: (++fdchangecnt)
```

According to the added debugging information:
fd_change: already set fdchanges and fdchangecnt values uccessfully；
fd_reify：the assignment is not performed.

In normal cases: a value is assigned to the loop in fd_reify；
In abnormal cases:
(1)thread2: fd_change set fdchangecnt value is 1;
(2)thread1:fd_reify run fdchangecnt -= changecnt， and changecnt value is 1, then fdchangecnt=0;
(3)thread1: next fd_reify execute ， fdchangecnt value is 0， loop will not update;
(4)thread2: af_unix_accept will not execute, so send_af_unix_string: cnt is always -1;
(5)CONNECTING will not accept,  the queue is not cleared, and the subsequent connection fails;

Please check whether this is the reason and how to fix it.

**Comment by stevegrubb** [comment]
Thanks for the analysis. Which version of the audit daemon are you testing with? There was a change from epoll to select backends in version 3.0.1.

**Comment by stillyger** [comment]
> Thanks for the analysis. Which version of the audit daemon are you testing with? There was a change from epoll to select backends in version 3.0.1.

Version is audit-3.0.1,  the function used is select_poll  and EV_SELECT_USE_FD_SET is 0

**Comment by stevegrubb** [comment]
Having looked into this a bit, what I think is the right answer is to only make calls libev from the main thread. Setting it up on init is OK because the event loop has not started. But once it does, all changes need to be made from the main thread. Modifying libev is not a good idea because subsequent updates to the next libev will be harder.

**Comment by stevegrubb** [comment]
Looking into this problem...besides adding a mutex (which solves races for 2 of many global variables) , there are 2 other options that come to mind.

1) Use ev_io_modify with EV_NONE and leave the listening descriptor open. Then reactivate it with another call to ev_io_modify with EV_READ.
2) Turn the plugin into an external plugin.

I think number 2 is the better option because it would take a significant refactoring to achieve 1. This could leave subtle bugs that may be non-obvious. By doing option 2, a whole lot of code can be removed. It could be drastically simplified. Additionally, an internal queue can be made so that auditd is free to move along faster.

**Comment by stevegrubb** [comment]
Commit 13c91c0 has the initial code to make an external af_unix plugin. If you have a chance to test it, I would appreciate feedback.

**Comment by stillyger** [comment]
> Commit [13c91c0](https://github.com/linux-audit/audit-userspace/commit/13c91c0273277846cdc8f97e0a103f2526ac7c24) has the initial code to make an external af_unix plugin. If you have a chance to test it, I would appreciate feedback.

Okay, thanks, I'll test this commit and feed back the results as soon as possible.

**Comment by stevegrubb** [comment]
Thanks. Note that there has been one other update past the commit mentioned because it was losing track of closed fd. Current code should be fine.

**Comment by stevegrubb** [comment]
I have changed the the conf files like you mentioned. Usually rpm overwrites them if they are unchanged. But I guess enabling the plugin changes it. Maybe I can continue parsing and then warn and switch internally.

For the second issue, I think changing the listen backlog to 1 might fix it. It is never intended to support more than 1 connection.

**Comment by stevegrubb** [comment]
I think that if the latest changes have prevented leaking socket fds, I should do a new release to get corrected code out to the community. I would like to do that later this week.

**Comment by stillyger** [comment]
I verified it with the latest commit，fds leakage still exists.
We created multiple socket clients parallel connect to `/var/run/audispd_events`，and close all client，then found leaks still exist.

I think If there is a new client connection, the socket fd `conn` value is overwritten, and only the last client value is retained. Old socket fds will no longer be processed. 
The listen backlog to 1 only restricts the data in the listening queue. But if accept4 called, the client is connected successfully,  then new client connections can cause leakage.

Check whether the old `conn` needs to be closed after the new `conn` is connected successfully.
